From: =?utf-8?q?Timo_R=C3=B6hling?= <roehling@debian.org>
Date: Thu, 13 Jun 2024 00:23:43 +0200
Subject: Fix test regressions for flaky test plugins

 A change in the behavior of the unittest teardown function caused a
 regression in plugins that re-ran tests. Namely python-flaky and
 pytest-rerunfailures. This patch was cherry-picked from an upstream
 PR.
Origin: upstream, https://github.com/pytest-dev/pytest/pull/12436
Bug: https://github.com/pytest-dev/pytest/issues/12424
---
 src/_pytest/unittest.py                                       |  2 +-
 .../plugins_integration/pytest_rerunfailures_integration.py   | 11 +++++++++++
 tox.ini                                                       |  2 +-
 3 files changed, 13 insertions(+), 2 deletions(-)
 create mode 100644 testing/plugins_integration/pytest_rerunfailures_integration.py

diff --git a/src/_pytest/unittest.py b/src/_pytest/unittest.py
index baca58b..2f3fd5b 100644
--- a/src/_pytest/unittest.py
+++ b/src/_pytest/unittest.py
@@ -216,7 +216,7 @@ class TestCaseFunction(Function):
             self._explicit_tearDown()
             self._explicit_tearDown = None
         self._obj = None
-        self._instance = None
+        del self._instance
         super().teardown()
 
     def startTest(self, testcase: "unittest.TestCase") -> None:
diff --git a/testing/plugins_integration/pytest_rerunfailures_integration.py b/testing/plugins_integration/pytest_rerunfailures_integration.py
new file mode 100644
index 0000000..9a13a32
--- /dev/null
+++ b/testing/plugins_integration/pytest_rerunfailures_integration.py
@@ -0,0 +1,11 @@
+import unittest
+
+
+class MyTestCase(unittest.TestCase):
+    first_time = True
+
+    def test_fail_the_first_time(self) -> None:
+        """Regression test for issue #12424."""
+        if self.first_time:
+            type(self).first_time = False
+            self.fail()
diff --git a/tox.ini b/tox.ini
index 30d3e68..e42c226 100644
--- a/tox.ini
+++ b/tox.ini
@@ -143,7 +143,7 @@ commands =
     pytest --cov=. simple_integration.py
     pytest --ds=django_settings simple_integration.py
     pytest --html=simple.html simple_integration.py
-    pytest --reruns 5 simple_integration.py
+    pytest --reruns 5 simple_integration.py pytest_rerunfailures_integration.py
     pytest pytest_anyio_integration.py
     pytest pytest_asyncio_integration.py
     pytest pytest_mock_integration.py
